changefeedccl: add job tracing support to changefeeds

asg0451 opened the pull request:
Adds support for job tracing to changefeeds. This
can be used as an additional debugging tool. The
data stored is summed span counts & times, per
job. The data can be accessed from the db console
like with other jobs.

Epic: none

Release note (ops change): Added job tracing support to changefeeds.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144412)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: add job tracing support to changefeeds

Adds support for job tracing to changefeeds. This
can be used as an additional debugging tool. The
data stored is summed span counts & times, per
job. The data can be accessed from the db console
like with other jobs.

Epic: none

Release note (ops change): Added job tracing support to changefeeds.

msbutler commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -318,6 +331,11 @@ func (ca *changeAggregator) Start(ctx context.Context) {
 	ctx = logtags.AddTag(ctx, changeAggregatorLogTag, nil /* value */)
 	ctx = ca.StartInternal(ctx, changeAggregatorProcName)
```
i believe you need to pass the aggregator to StartInternal, similar to other bulk jobs.

msbutler commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -758,6 +776,15 @@ func (ca *changeAggregator) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMet
+		select {
+		case <-ca.aggTimer.C:
+			ca.aggTimer.Read = true
+			ca.aggTimer.Reset(1 * time.Second)
```
could be worth following convention to reset this timer every 15 seconds, so it intercepts with next() less frequently

msbutler commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1297,9 +1335,16 @@ func (cf *changeFrontier) Start(ctx context.Context) {
 		ctx = logtags.AddTag(ctx, "job", cf.spec.JobID)
 	}
 	ctx = logtags.AddTag(ctx, changeFrontierLogTag, nil /* value */)
+
```
NB: PCR only traces the ingestion processor, where the vast majority of work occurs, and not its frontier processor. It may not be neccessary to trace the changefeed frontier processor, but you have more context.

msbutler commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1297,9 +1335,16 @@ func (cf *changeFrontier) Start(ctx context.Context) {
+
 	// StartInternal called at the beginning of the function because there are
 	// early returns if errors are detected.
 	ctx = cf.StartInternal(ctx, changeFrontierProcName)
```
ditto to passing the aggregator to the StartInternal

msbutler commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -1655,7 +1690,20 @@ func (b *changefeedResumer) OnFailOrCancel(
+		aggStatsCopy = b.mu.perNodeAggregatorStats.DeepCopy()
+	}()
+
+	if err := bulk.FlushTracingAggregatorStats(ctx, b.job.ID(),
```
nit: just `return bulk.FlushTracing...()`?

msbutler reviewed (approved):
Nice! are you still getting weird shutdown deadlocks? mind posting a screenshot of the new trace?

asg0451 commented:
tyftr

> Nice! are you still getting weird shutdown deadlocks? mind posting a screenshot of the new trace?

@msbutler nope, fixed that. however the issue where the trace text has no tabs/hierarchy is still there. i think you indicated that could be a regression elsewhere in the tracing system but just wanna make sure it's not me doing smth wrong

msbutler reviewed (approved):
We see the lack of hierarchy in backup traces too, so i don't think you're holding anything incorrectly
